src/storage/txn/commands/check_secondary_locks.rs

```diff
@@ -54,6 +54,85 @@ enum SecondaryLockStatus {
     RolledBack,
 }
 
+// The returned `bool` indicates whether the rollback record should be written,
+// it should be true if and only if the txn commit record is not found, thus
+// a rollback record would be written later.
+fn check_determined_txn_status<S: Snapshot>(
+    reader: &mut ReaderWithStats<'_, S>,
+    key: &Key,
+) -> Result<(SecondaryLockStatus, bool, Option<OverlappedWrite>)> {
+    match reader.get_txn_commit_record(key)? {
+        TxnCommitRecord::SingleRecord { commit_ts, write } => {
+            let status = if write.write_type != WriteType::Rollback {
+                SecondaryLockStatus::Committed(commit_ts)
+            } else {
+                SecondaryLockStatus::RolledBack
+            };
+            // We needn't write a rollback once there is a write record for it:
+            // If it's a committed record, it cannot be changed.
+            // If it's a rollback record, it either comes from another
+            // check_secondary_lock (thus protected) or the client stops commit
+            // actively. So we don't need to make it protected again.
+            Ok((status, false, None))
+        }
+        TxnCommitRecord::OverlappedRollback { .. } => {
+            Ok((SecondaryLockStatus::RolledBack, false, None))
+        }
+        TxnCommitRecord::None { overlapped_write } => {
+            Ok((SecondaryLockStatus::RolledBack, true, overlapped_write))
+        }
+    }
+}
+
+fn check_status_from_lock<S: Snapshot>(
+    txn: &mut MvccTxn,
+    reader: &mut ReaderWithStats<'_, S>,
+    lock: Lock,
+    key: &Key,
+    region_id: u64,
+) -> Result<(
+    SecondaryLockStatus,
+    bool,
+    Option<OverlappedWrite>,
+    Option<ReleasedLock>,
+)> {
+    let mut overlapped_write = None;
+    if lock.is_pessimistic_lock_with_conflict() {
+        assert!(lock.is_pessimistic_lock());
+        let (status, need_rollback, rollback_overlapped_write) =
+            check_determined_txn_status(reader, key)?;
+        // If there exists commit or rollback record, the pessimistic lock is stale, in
+        // this case the returned need_rollback is false.
+        if !need_rollback {
+            let released_lock = txn.unlock_key(key.clone(), true, TimeStamp::zero());
+            return Ok((
+                status,
+                need_rollback,
+                rollback_overlapped_write,
+                released_lock,
+            ));
+        }
+        overlapped_write = rollback_overlapped_write;
+    }
+
+    if lock.is_pessimistic_lock() {
+        let released_lock = txn.unlock_key(key.clone(), true, TimeStamp::zero());
+        let overlapped_write_res = if lock.is_pessimistic_lock_with_conflict() {
+            overlapped_write
```

Comment:
Consider comment about that `overlapped_write` is already loaded if the above lock_with_conflcit related code is executed so that we can use it directly.